lib/app/modules/onc_program/onc_ehr_launch_sequence.rb

```diff
@@ -17,63 +17,69 @@ class OncEHRLaunchSequence < EHRLaunchSequence
-        'Condition',
-        'Device',
-        'DiagnosticReport',
-        'DocumentReference',
-        'Encounter',
-        'ExplanationOfBenefit',
-        'Goal',
-        'Immunization',
-        'Medication',
-        'MedicationDispense',
-        'MedicationStatement',
-        'MedicationOrder',
-        'Observation',
-        'Procedure',
-        'DocumentReference',
-        'Provenance'
-      ]
-
-      test 'Scopes enabling user-level access with OpenID Connect and Refresh Token present' do
+      def valid_resource_types
+        [
+          '*',
+          'Patient',
+          'AllergyIntolerance',
+          'CarePlan',
+          'Condition',
+          'Device',
+          'DiagnosticReport',
+          'DocumentReference',
+          'Encounter',
+          'ExplanationOfBenefit',
+          'Goal',
+          'Immunization',
+          'Medication',
+          'MedicationDispense',
+          'MedicationStatement',
+          'MedicationOrder',
+          'Observation',
+          'Procedure',
+          'DocumentReference',
+          'Provenance'
+        ]
+      end
+
+      test :onc_scopes do
         metadata do
           id '11'
+          name 'Scopes enabling user-level access with OpenID Connect and Refresh Token present'
           link 'http://www.hl7.org/fhir/smart-app-launch/scopes-and-launch-context/index.html#quick-start'
           description %(
             The scopes being input must follow the guidelines specified in the smart-app-launch guide
           )
         end
+        required_scopes = ['openid', 'fhirUser', 'launch', 'offline_access']
         scopes = @instance.scopes.split(' ')
 
-        assert scopes.include?('openid'), 'Scope did not include "openid"'
-        scopes.delete('openid')
-        assert scopes.include?('fhirUser'), 'Scope did not include "fhirUser"'
-        scopes.delete('fhirUser')
-        assert scopes.include?('launch'), 'Scope did not include "launch"'
-        scopes.delete('launch')
-        assert scopes.include?('offline_access'), 'Scope did not include "offline_access"'
-        scopes.delete('offline_access')
+        missing_scopes = required_scopes - scopes
+        assert missing_scopes.empty?, "Required scopes missing: #{missing_scopes.join(', ')}"
 
+        scopes -= required_scopes
         # Other 'okay' scopes
         scopes.delete('online_access')
 
         user_scope_found = false
 
         scopes.each do |scope|
+          bad_format_message = "Scope '#{scope}' does not follow the format: user/[ resource | * ].[ read | * ]"
           scope_pieces = scope.split('/')
-          assert scope_pieces.count == 2, "Scope '#{scope}' does not follow the format: user/[ resource | * ].[ read | * ]"
-          assert scope_pieces[0] == 'user', "Scope '#{scope}' does not follow the format: user/[ resource | * ].[ read | * ]"
+
+          assert scope_pieces.count == 2, bad_format_message
+          assert scope_pieces[0] == 'user', bad_format_message
```

Comment:
Do we really want to fail if a user-level rather than patient-level scope is present (and vice-versa for the other sequence)?

Reply:
If we are saying "prove you can support a user scope and not just patient scopes" then I think its reasonable.  But I think this would require more off-line discussion.